app/channel.go

```diff
@@ -716,6 +722,137 @@ func (a *App) GetSchemeRolesForChannel(channelId string) (string, string, string
 	return model.CHANNEL_GUEST_ROLE_ID, model.CHANNEL_USER_ROLE_ID, model.CHANNEL_ADMIN_ROLE_ID, nil
 }
 
+func (a *App) GetChannelModerationsForChannel(channel *model.Channel) ([]*model.ChannelModeration, *model.AppError) {
+	guestRoleName, memberRoleName, _, _ := a.GetSchemeRolesForChannel(channel.Id)
+	memberRole, err := a.GetRoleByName(memberRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	guestRole, err := a.GetRoleByName(guestRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	inheritedGuestRoleName, inheritedMemberRoleName, _, _ := a.GetTeamSchemeRolesForChannel(channel.TeamId)
+	inheritedMemberRole, err := a.GetRoleByName(inheritedMemberRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	inheritedGuestRole, err := a.GetRoleByName(inheritedGuestRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	return buildChannelModerations(memberRole, guestRole, inheritedMemberRole, inheritedGuestRole), nil
+}
+
+func (a *App) PatchChannelModerationsForChannel(channel *model.Channel, channelModerationsPatch []*model.ChannelModerationPatch) ([]*model.ChannelModeration, *model.AppError) {
+	inheritedGuestRoleName, inheritedMemberRoleName, _, _ := a.GetTeamSchemeRolesForChannel(channel.TeamId)
+	inheritedMemberRole, err := a.GetRoleByName(inheritedMemberRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	inheritedGuestRole, err := a.GetRoleByName(inheritedGuestRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	inheritedMemberPermissions := GetChannelModeratedPermissions(inheritedMemberRole.Permissions)
+	inheritedGuestPermissions := GetChannelModeratedPermissions(inheritedGuestRole.Permissions)
+
+	for _, moderationPatch := range channelModerationsPatch {
+		if moderationPatch.Roles.Members != nil && *moderationPatch.Roles.Members && !inheritedMemberPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+		if moderationPatch.Roles.Guests != nil && *moderationPatch.Roles.Guests && !inheritedGuestPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+	}
+
+	// Channel has no scheme so create one
+	if channel.SchemeId == nil || len(*channel.SchemeId) == 0 {
+		scheme := &model.Scheme{
+			Name:        model.NewId(),
+			DisplayName: model.NewId(),
+			Scope:       model.SCHEME_SCOPE_CHANNEL,
+		}
+		if scheme, err = a.CreateScheme(scheme); err != nil {
+			return nil, err
+		}
+		channel.SchemeId = &scheme.Id
+		a.UpdateChannelScheme(channel)
```

Comment:
Do you want to check for an error here?

Reply:
I abstracted this into a different method and handled errors properly